### DISCUSSION TITLE
JSON Example

### DISCUSSION THREAD

**Discussion by yesoreyeram** [grafana/grafana-infinity-datasource#34] [Show and tell]
## JSON URL

Below example shows about fetching data from JSON URL.

URL : https://jsonplaceholder.typicode.com/users

![image](https://user-images.githubusercontent.com/153843/108413678-34314c80-7223-11eb-9cce-603134ec2d48.png)

In the above example, the data in the URL is array. So no need to configure any additional fields except url in the panel.

## JSON URL - data in a json property

URL : https://thingspeak.com/channels/38629/feed.json

![image](https://user-images.githubusercontent.com/153843/108415043-de5da400-7224-11eb-9295-d49fcc18464a.png)

In the above example, data is in the `feeds` property. So the same is specified as root / rows field. But still our plugin doesn't know anything about the fields or it's types. So we are going to add the columns to make it more meaningful. 

We are adding columns and defining their types as shown below.

![image](https://user-images.githubusercontent.com/153843/108427049-7dd66300-7234-11eb-8d27-cec50945a66c.png)

## JSON Data without time field

URL : https://gist.githubusercontent.com/yesoreyeram/2433ce69862f452b9d0460c947ee191f/raw/f8200a62b68a096792578efd5e3c72fdc5d99d98/population.json

![image](https://user-images.githubusercontent.com/153843/108415716-cdf9f900-7225-11eb-8e0d-5d767104a080.png)

In the above example, we are visualising a json data without time field. Our JSON  has only two fields aka `country` and `population`.  So we asked the plugin to add a dummy time field to the data so that we can visualize them in any of the grafana's stock panel. If you closely look at the image above, you can see we specified 'format' as **timeseries**.  

For reference, JSON data from the URL is given below

```json
[
  { "country": "india", "population": 300 },
  { "country": "usa", "population": 200 },
  { "country": "uk", "population": 150 },
  { "country": "china", "population": 400 }
]
```

**Comment by charishma1** [comment]
Hi,I am using the URL to fetch json data from REST API.
i am trying to use dynamic multi value dashboard variables in the middile of the URL,when i select the multiple vaulues of the variable,am able to get the json data in multiple panels considering the panel repetiton option is enabled.
But i am trying to fetch the json data of multiple urls n the same panel,which is failing.
Any help would be appreciated.

**Reply by yesoreyeram** [reply]
No. You can't repeat variables inside the same panel. By design it is not possible.

**Reply by charishma1** [reply]
Thank you for the reply .
Then is there any other way i can combine the data coming in repeated panels ?
or is there a possibility that this can be added as new feature in the upcoming releases ?

**Comment by klindsay-dowlerkarn** [comment]
Hi,  Is there something in this datasource that would stop the URL from working if the endpoint was on the same local network as my on premise Grafana server but was not resolvable to the public internet?  I have a private webserver on the same private network as my Grafana server but when I try to use this Infinity json datasource with my endpoint all I get is a 500 error.  I can access my endpoint via the browser without issue and have even gone so far as to copy the json result set exactly from your sample URL in my endpoint, but all I ever get is a 500 error.  Any thoughts as to why my endpoint can't be resolved by this datasource?

**Reply by yesoreyeram** [reply]
No. That should work. You have to provide the address that the machine grafana runs understands. ie, curl to the url from the machine grafana runs and make sure to use the same url

**Reply by klindsay-dowlerkarn** [reply]
Curl does work from the Grafana server with one caveat, I have to use curl -k to ignore a certificate error.  The Grafana server recognizes my wildcard certificate just fine but I'm guessing whatever is incorrect for the curl setup is also affecting the Infinity data source.

**Reply by klindsay-dowlerkarn** [reply]
For anyone else who runs into this issue, you need to make sure the base server (in my case ubuntu) is setup properly to use your certificate not just the Grafana server.  Once you can curl to the desired URL from the Grafana server without getting a certificate error, restart the Grafana service and the Infinity data source will work (at least it did for me).  Thanks for the help yesoreyeram.

**Reply by roylisanudin** [reply]
data [
	{
		id : "222",
		class : "Platinum",
		area : "Aceh",
		
		avail : [
			- {
				av_id : "112",
				Stat : "red",
				sla : 11
			} ,
			- {
				av_id : "11",
				Stat : "blue"
				sla : 1
			}
		]
		}
	]


Hello, how can i call value from avail in rows/root (URL) ? 
I need a way without going through the "Inline" source, I want it auto via URL. Need answer please

**Comment by asktushar** [comment]
Hello, I was hoping if you can create some examples for a scenario where your JSON has an array field and how would you read the data from those data fields. I have also opened a [Discussion](https://github.com/yesoreyeram/grafana-infinity-datasource/discussions/206) on the same topic.

**Comment by mahmutefe** [comment]
Is it possible to create an alert via the infinity plugin? I am trying and receiving the below error.
API: https://thingspeak.com/channels/38629/feed.json
![image](https://user-images.githubusercontent.com/5229401/158916848-16e85e89-d283-4f90-bb91-37b3d7998c6c.png)

**Reply by yesoreyeram** [reply]
No. Alerts not possible at the moment

**Reply by yesoreyeram** [reply]
1.0.0-beta.1 is now **[available](https://github.com/yesoreyeram/grafana-infinity-datasource/releases/tag/v1.0.0-beta.1)** with alerting support. 

With JSON query type, one need to choose `backend` as a parser to get the alerts working. Example shown below for setting up alert.

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/153843/188924894-4b1f3d3d-2857-484e-916c-d09cb8d4dd5c.png">

Example API used : https://thingspeak.com/channels/38629/feed.json

**Reply by nkroeker** [reply]
Is alerting broken again? I have been trying to create an alert with the backend parser and get only this error no matter what I do.
Tried on plugin version  2.9.5 and 3.0 beta

[sse.dataQueryError] failed to execute query [A]: error un-marshaling the query: error while parsing the query json. json: cannot unmarshal bool into Go struct field Query.range of type string

![image](https://github.com/user-attachments/assets/0cb0d2bf-9788-4357-b07b-2923363d1aac)

**Reply by Dexter-03** [reply]
The bug seems to lay in the google-sheets Type.
Workaround:
- Switch the Type from JSON to Google-Sheets
- Change the range from true to "123" for instance
- Switch back to JSON and configure everything as you like.
For me it works this way. Still a bug that should be solved.

**Comment by dharmeshprataps** [comment]
Hey currently I am on trial and I want to access some data from my company fql database. I am able using to access data using a curl (POST) request but I am not able to get the data using post command in infinity dashboard and when i got into the detail i can see the headers I am setting their values are automatically becoming xxxxxxx. Is there any particular reason for that. Can't find any documentation regarding making POST queries in infinity dashboard. Please help at the earliest.

**Reply by yosiasz** [reply]
Would this help?

https://sriramajeyam.com/grafana-infinity-datasource/wiki/graphql

**Reply by yosiasz** [reply]
I was answering @dharmeshprataps

**Comment by franciscogv-itx** [comment]
Hi,

Currently trying to filter an array based on the value of another (multivalued) variable by using regex but so far haven't had any luck.

I prepared an example by using your same data in case you can lend me a hand

If i filter the array by simply using an equality condition (==) it works as expected 
![image](https://user-images.githubusercontent.com/82811517/181745010-6e64640d-4828-458d-b340-1d911d5c3b53.png)

but as soon as I change it to a regex it looks like it completely ignores the filter and simply returns all possible results
![image](https://user-images.githubusercontent.com/82811517/181745300-38d1f9d6-dbf7-4a5e-9a52-3cdc9aae296b.png)

Haven't found much info on this and uql is not an option (as far as I know) for variables so is there a way to solve this?

Thanks in advance

**Reply by KravchenkoAlex** [reply]
i faced with the same issue, did find solution. Did you solve it?

**Reply by yosiasz** [reply]
Check this out with UQL
https://try.jsonata.org/oS2PoQvLa

**Reply by KravchenkoAlex** [reply]
> Check this out with UQL https://try.jsonata.org/oS2PoQvLa

okey, uql might solve my issue, but didn't get how check it on the site for jsonNata

**Reply by KravchenkoAlex** [reply]
Okey, i got it. We can use jsonNata in UQL. But anyway didn't get. This selector works for me result[name="May_23_5:20 PM"].id
How use it with UQL?
![image](https://github.com/yesoreyeram/grafana-infinity-datasource/assets/19817080/c23749d5-e1f9-4fa0-9558-2cdddac7e983)
didn't work. Im sure i did something wrong

**Reply by yosiasz** [reply]
What have you tried?

**Reply by KravchenkoAlex** [reply]
Okey, so jsonata selector works result[name="May_23_5:20 PM"].id on site with my json 
I've tried to put this selector to uql
![image](https://github.com/yesoreyeram/grafana-infinity-datasource/assets/19817080/72f54e55-bbce-4c10-a683-7979b62896bc)
doesn't work.
My goal is: I've already created variable in grafana variables which takes list of all names. And now im trying to create second variable which get the id according to chosen "report_name"
![image](https://github.com/yesoreyeram/grafana-infinity-datasource/assets/19817080/46e5c5b5-952a-4c22-9166-7788d5dd3545)
Something like that. Did you get my point?

**Reply by KravchenkoAlex** [reply]
@yosiasz i did it 
parse-json 
| jsonata "*[name='${report_name}']" 
| project "id"

**Comment by rajeshchella2887** [comment]
Is possible to compare two json files and store the results into a dashboard.

![image](https://user-images.githubusercontent.com/111831451/197155470-81524e89-0346-419c-8b1f-3bdc35a2592c.png)

**Reply by rajeshchella2887** [reply]
Updating above request

I am Trying to compare two json api urls like url:1 and url:2 and store the output into a dashboard.

url1 contains value : username , userid, status

url2 contains value : userlocation , userprofile

In final ouput i expect the results in dashboard combined both urls datas. 

dashboard  result : username , userid, status, userlocation , userprofile

(or)

Another query like combine both api urls url:1 and url:2 into a single GET request ?

![image](https://user-images.githubusercontent.com/111831451/200280316-0df70c3d-1885-4427-80a1-97bbf0857bf4.png)

**Reply by yesoreyeram** [reply]
choose the right tool for the job.. I am not sure what you are trying to achieve by comparing two results in grafana. Seems very unique to your problem.

**Comment by sirAlexander** [comment]
Hi @yesoreyeram would you be so kind as to point me to resources where I can learn UQL

also any infinity plugin pivot table like examples would be greatly appreciated.

Thanks

**Reply by yesoreyeram** [reply]
learn about UQL in
* https://yesoreyeram.github.io/grafana-infinity-datasource/wiki/uql
* https://github.com/yesoreyeram/uql ( test cases )

**Reply by sirAlexander** [reply]
Thank you!

**Comment by Lroy** [comment]
I am having great success in using Infinity for fetching single-series data from my custom Json data service.  This is a very nice datasource!

**Reply by Lroy** [reply]
I am, however, having some difficulty in figuring out how to format my Json for more than 1 series.  Can you provide an example of how to format the Json in order to plot two values from two different data series?

For example, I have server 1 and server 2.  Each has a CPU load field.  I would like to plot the last 15 minutes of CPU load for each server on the same time series panel.  I cannot seem to get my mind around how to format the Json.

**Comment by shahzi176** [comment]
Hello, I need help, I am using grafana 7.0.0 and infinity as datasource, I have an API which I want to show in tables, But problem here is API structure, the information I am trying to utilize is under data and than there is further blocks. how can I display the information in the casecading block of data?

**Reply by wmashal** [reply]
I think infinity just working with Grafana >=8.4.7

**Comment by wmashal** [comment]
Hi

Hi 
Is there any example shows how I can group by string column like the following

![image](https://github.com/yesoreyeram/grafana-infinity-datasource/assets/11045408/fcc511c6-27ec-4762-b78d-9124961650d8)

Json sample 

```json


feature-toggles: [
{
name: "cc-limit-user-bindings",
namespace: "feature-toggle",
creationTimestamp: "2023-07-02T05:32:06Z",
modefiedTimestamp: "2023-07-02T05:32:06Z",
enabled: false,
landscape: "staging",
componentName: "core-controllers",
createdBy: "Shachar Malachi",
version: 6,
strategy: {
name: "default"
}
},
{
name: "cim-publish-to-root-governance",
namespace: "feature-toggle",
creationTimestamp: "2023-07-02T05:32:06Z",
modefiedTimestamp: "2023-07-02T05:32:06Z",
enabled: true,
landscape: "staging",
componentName: "commercial-integration-manager",
createdBy: "Gilad Keinan",
version: 6,
strategy: {
name: "default"
}
},
{
name: "test-feature-toggle-lock",
namespace: "feature-toggle",
creationTimestamp: "2023-07-02T05:32:06Z",
modefiedTimestamp: "2023-07-02T05:32:06Z",
enabled: null,
landscape: "canary",
componentName: "some-test-ft-scenario",
createdBy: "Nati Gavriel",
version: 1,
strategy: {
name: "default"
}
},
{
name: "ua-restrict-initial-admins",
namespace: "feature-toggle",
creationTimestamp: "2023-07-02T05:32:06Z",
modefiedTimestamp: "2023-07-02T05:32:06Z",
enabled: true,
landscape: "canary",
componentName: "some-test-ft-scenario",
createdBy: "Aviv Guy",
version: 2,
strategy: {
name: "default"
}
}
]


```

**Reply by wmashal** [reply]
@yesoreyeram any thoughts ?

**Reply by ravinayag** [reply]
I do have similar json outputs, and want by `group by` version and their counts.

```
{
"peerInfo":[
{
"peerId": "EiBZ3p0CtvTG0589kN+vT/nDXIpdr3WEU5b0kJ8kxelWiQ==",
"multiaddrs":[""],
"maxFrame": "11692",
"timestamp": "1704772793821",
"version": "AQIB",
"signature": "/77G1fAjB85pZrwBjSjfIgxJafXw1kaxs6WyRXY5uVrSRlukIPziFqF7v/RbRSuP5CRlh7UVId+AkiNUx+RdNt+tg12vHjboJZZ8R4R+vHhNYP6KcNNWjT8mxBvjp4oL/3uCr/kDx9DVIfz/4lDV4DkA",
"publicKey": "XE6XpWupJp4g/wwvdpS2QhSLRivO5jb96gmjFS1nyw8rbiJBkkRVW7Ofm8JNeN+4VfCvIArxB3cA"
},
{
"peerId": "EiBtOhb9i8DyOfFipffoUu0STTSGq470VMnzUwbdBeIUzw==",
"multiaddrs":[""],
"maxFrame": "11849",
"timestamp": "1704772677195",
"version": "AQIB",
"signature": "R99N/K89nhJMO3uM59x8chtkC9OzijH3wtHMbWtv0eQxO0lXD4Gs8j8WFP7b8J+fEoj08BoR0TUAzXzl/En+xjQO5JzXoAAnRvA/mH72RHv2EB1ftLDFDonJC3+wKE7X8MN6CEF2Qi8Bw7Mx66f1zwEA",
"publicKey": "1jtT1rL1CRgQf2TLNJG0SvPZx5MMUPDw7qUXv7YxYYKy7xdCDWdhma6kAVDA4PmIGlq/OyhHNWAA"
},
{"peerId": "EiCLibPSoQAjMaAl0Xtf8bw8gWeJ/9bnk7PSu42LZ3OPuA==", "multiaddrs":["" ], "maxFrame": "16121",…},
```

**Comment by smtpl-iiotian** [comment]
I came across an error where the grafana table-old panel displayed "Query data error". 
Error observed while backend parser was selected and the time range from 23:00:00 to 07:00:00 (i.e, shift 3). For other time ranges which are in a single day (i.e, shift1, shift2) backend parser works same as default. Default parser works for all time ranges but backend parser fails for shift3 time range. 
As seen in /var/log/syslog first line of error message shows
lvl=dbug msg="panic: interface conversion: interface {} is *string, not *float64"

Trying to understand where lies the issue. 
Kindly, guide.

**Comment by lakshay2711** [comment]
@yesoreyeram I need to trigger alert using infinity plugin, I am using parser as Backend, in my case I am querying from Elasticsearch index and it gives me the result when I use hits.hits in the "Rows/Root" field in create alert, but it is failing when I use "hits.hits._source" Actually requirement is like I need to access the "_source" property inside "hits.hits" of every document and add one more field dynamically and then query based on a filter condition for two properties of the _source object, can you please let me know how can I achieve this.

Please consider below sample response from Elasticsearch:

```
"hits" : {
    "total" : {
      "value" : 10
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_source" : {
          "orderCreatedDate" : "2023-10-15 13:38:04",
           ........
        }
      },
      {
        "_source" : {
          "orderCreatedDate" : "2023-10-18 13:38:04",
           ........
        }
      }
 
```

**Reply by yesoreyeram** [reply]
You just want `hits.hits` as root selector.

**Comment by Blecao** [comment]
Can you use this method to retrieve the http status code instead of the data itself or it isnt posible
Thanks in advance

**Reply by yesoreyeram** [reply]
Nope.. currently it is not possible. What you need is synthetic monitoring feature of grafana cloud

**Reply by Blecao** [reply]
Thank you for your quick awnser 
Best regards

**Comment by sINGH1996202** [comment]
My API output is like the following 

```json
{
  "properties": {
    "nextLink": null,
    "columns": [
      {
        "name": "Cost",
        "type": "Number"
      },
      {
        "name": "UsageDate",
        "type": "Number"
      },
      {
        "name": "Currency",
        "type": "String"
      }
    ],
    "rows": [
      [232323, 20240101, "USD"],
      [232323, 20240102, "USD"]
    ]
  }
}

```
 How do I separate the 3 values inside "rows"?

**Comment by cpgarcia518** [comment]
In my case the infinity plugin does not work when I request the data from a local server (localhost:1415).
![image](https://github.com/grafana/grafana-infinity-datasource/assets/50991879/2fe3ab11-82f8-4146-b375-fff8e0e323c4)

**Reply by yesoreyeram** [reply]
Does grafana running on the same server? If not, grafana doesn't know how to connect to your API. If you are using docker compose kind of thing, you need to use the service name instead of localhost.

**Reply by pablorotten** [reply]
Hola @cpgarcia518 !!!

The error message `dial tcp 127.0.0.1` smells like you're running Grafana in a docker and you are trying to access to your local server web app, which is outside of the Grafana docker. So if you put `localhost` in Grafana, will try to access to the localhost inside the docker not the one of you computer.

Make sure you use `--network="host"` in your docker run command, then `127.0.0.1` in your docker container will point to your docker host. https://stackoverflow.com/a/24326540

**Reply by cpgarcia518** [reply]
Thank you so much @yesoreyeram and @pablorotten for your support, it is working fine. Specifically I have defined notework_mode = "host" in order to share the container network configuration with the host machine.

**Comment by rajeshvardhan** [comment]
```json
{
  "_class": "hudson.model.ComputerSet",
  "busyExecutors": 11,
  "computer": [
    {
      "_class": "hudson.model.Hudson$MasterComputer",
      "actions": [
        {
          
        }
      ],
      "assignedLabels": [
        {
          "name": "master"
        }
      ],
      "description": "the Jenkins controller's built-in node",
      "displayName": "Built-In Node",
      "executors": [
        {
          
        }
      ],
      "icon": "symbol-computer",
      "iconClassName": "symbol-computer",
      "idle": false,
      "jnlpAgent": false,
      "launchSupported": true,
      "loadStatistics": {
        "_class": "hudson.model.Label$1"
      },
      "manualLaunchAllowed": true,
      "monitorData": {
        "hudson.node_monitors.SwapSpaceMonitor": {
          "_class": "hudson.node_monitors.SwapSpaceMonitor$MemoryUsage2",
          "availablePhysicalMemory": 89718784,
          "availableSwapSpace": 8182034432,
          "totalPhysicalMemory": 2171736064,
          "totalSwapSpace": 8455712768
        },
        "hudson.node_monitors.TemporarySpaceMonitor": {
          "_class": "hudson.node_monitors.DiskSpaceMonitorDescriptor$DiskSpace",
          "timestamp": 1707968517427,
          "path": "/tmp",
          "size": 15627452416
        },
        "hudson.node_monitors.DiskSpaceMonitor": {
          "_class": "hudson.node_monitors.DiskSpaceMonitorDescriptor$DiskSpace",
          "timestamp": 1707968517302,
          "path": "/var/lib/jenkins",
          "size": 15627452416
        },
        "hudson.node_monitors.ArchitectureMonitor": "Linux (amd64)",
        "hudson.node_monitors.ResponseTimeMonitor": {
          "_class": "hudson.node_monitors.ResponseTimeMonitor$Data",
          "timestamp": 1707968517313,
          "average": 0
        },
        "hudson.node_monitors.ClockMonitor": {
          "_class": "hudson.util.ClockDifference",
          "diff": 0
        }
      },
      "numExecutors": 1,
      "offline": false,
      "offlineCause": null,
      "offlineCauseReason": "",
      "oneOffExecutors": [
        {
          
        },
        {
          
        },
        {
          
        },
        {
          
        }
      ],
      "temporarilyOffline": false
    },
    {
      "_class": "hudson.slaves.SlaveComputer",
      "actions": [
        {
          "_class": "hudson.plugins.jobConfigHistory.ComputerConfigHistoryAction"
        },
        {
          
        },
        {
          
        }
      ],
      "assignedLabels": [
        {
          "name": "aquarover"
        },
        {
          "name": "linux84"
        }
      ],
      "description": "",
      "displayName": "aquarover",
      "executors": [
        {
          
        }
      ],
      "icon": "symbol-computer-offline",
      "iconClassName": "symbol-computer-offline",
      "idle": true,
      "jnlpAgent": true,
      "launchSupported": false,
      "loadStatistics": {
        "_class": "hudson.model.Label$1"
      },
      "manualLaunchAllowed": true,
      "monitorData": {
        "hudson.node_monitors.SwapSpaceMonitor": null,
        "hudson.node_monitors.TemporarySpaceMonitor": null,
        "hudson.node_monitors.DiskSpaceMonitor": null,
        "hudson.node_monitors.ArchitectureMonitor": null,
        "hudson.node_monitors.ResponseTimeMonitor": null,
        "hudson.node_monitors.ClockMonitor": null
      },
      "numExecutors": 1,
      "offline": true,
      "offlineCause": null,
      "offlineCauseReason": "",
      "oneOffExecutors": [
        
      ],
      "temporarilyOffline": false,
      "absoluteRemotePath": null
    }
  ]
}
```


Here is my json data. I want to access under computer , under monitoring data, availablePhysicalMemory. Can you please help me how can I acheive this

**Comment by dr-wut** [comment]
@yesoreyeram 
Good afternoon
Please tell me, do you have examples of json that are correctly displayed as traces (traces visualization)?
Those examples that are on the grafana website do not work =(

**Reply by yesoreyeram** [reply]
See if this example helps https://play.grafana.org/goto/r1G082oIg?orgId=1

Also document available in https://grafana.com/docs/plugins/yesoreyeram-infinity-datasource/latest/references/display-options/tracing/

**Reply by yesoreyeram** [reply]
The above example is for csv. But the same can be applied to JSON as well.

**Reply by dr-wut** [reply]
Great! Thank you very much for your answer. I managed to make a working json, but I don’t understand how to add Span Attributes and Resource Attributes. Please tell me how should I deal with this?

**Reply by yesoreyeram** [reply]
<img width="1792" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/bc677a3e-b7a6-4fd5-8db9-28526c1ae294">

I don't think span attributes/ resource attributes are possible right now. Can you create a issue/feature request for this? So this can be considered in next versions.

**Comment by baddar99** [comment]
Hi, i have a question that my Api request is dependent on two steps

1. get all the end points in form of json object e.g.  `[
  {
    "active": true, 
    "path": "get/webServices/weather/",
    "startedAt": "2024-03-04T15:19:09.711094Z",
    "endsAt": null,
    "updatedAt": "2024-03-18T08:48:43.790188Z"
  },
{
    "active": true, 
    "path": "get/webServices/temperautre/",
    "startedAt": "2024-03-04T15:19:09.711094Z",
    "endsAt": null,
    "updatedAt": "2024-03-18T08:48:43.790188Z"
  }
]`
2. the real data will come from the "path" ends points 
can you help how can i achieve this thing?

**Reply by yesoreyeram** [reply]
Have your first query as variable query and use the value in actual query URL.

**Reply by baddar99** [reply]
@yesoreyeram thanks for the reply, i have to ask one more thing that can we extend this to more than 2 levels?
if yes, how we can do that?

**Comment by xcatza** [comment]
Hi, I have a question regarding variables. I've used inline JSON to set the following:

`[{
"Site": "Melbourne Office",
"State": "VIC",
"Address": "Palmer"
},
{
"Site": "Sydney Office",
"State": "NSW",
"Address": "Surry Hills"
},
{
"Site": "Brisbane Office",
"State": "QLD",
"Address": "Milton"
}]`

And within the dashboards, I've set a panel to show these as columns, depending on which site is selected. The query itself also uses an inline CSV format:

`"Site","Address","State"
${sitesearch:text},${sitesearch:Address},${sitesearch:State}
`
However, this doesn't seem to work and would cause all columns to be the state. I've tried setting the columns in the variable menu as mentioned in the doc note [here](https://grafana.com/docs/plugins/yesoreyeram-infinity-datasource/latest/variables/variables/#:~:text=Note%3A%20It%20is,column%20for%20variable.)

![image](https://github.com/grafana/grafana-infinity-datasource/assets/42014474/c9ecc71f-b5fc-44a2-9bff-55f057b2f067)

But that doesn't seem to change much. Is there something i'm missing or is this not possible?

**Comment by ashwaqrehman** [comment]
<img width="932" alt="Screenshot 2024-04-04 at 4 33 11 PM" src="https://github.com/grafana/grafana-infinity-datasource/assets/156284074/5ace3e9d-85bf-4677-8c7d-c99815aedb4a">

Hi, I'm unable to set the alerts even though I have choose the backend parser, Can anyone guide me in achieving this thing? I want to create alerts on JSON data called through API

**Comment by dmitrynovik** [comment]
I have a JSON I get from an HTTP endpoint without a timestamp, and I want to add an artificial timestamp as the local time I do this HTTP GET. Is that possible, and how?

I was looking into the "Transform Data" but could not figure out.

**Reply by yesoreyeram** [reply]
If you are using backend parser, You can get help from JSONata and grafana time macro

```js
$map($,function($item){
   $merge([$item,{ "ts" : ${__to} }])
})
```

<img width="1721" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/7fcfb9f8-95bf-4658-a793-7c3386e9f183">


If you use UQL parser, again from the grafana time macro

```kusto
parse-json
| extend "ts"=unixtime_milliseconds_todatetime(${__to})
```

<img width="1728" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/dd508979-837d-47be-9ef3-457f1d7dbbab">

**Comment by arnebracke** [comment]
How to use pagination in zero-based API's?
I can't reset the initial page number to 0 -> it is automatically overwritten by its default being 1:

![image](https://github.com/grafana/grafana-infinity-datasource/assets/173165218/6e9d0567-cb5b-40b4-b13f-4498f93a82ee)

Is there a workaround at this moment?

**Comment by Nikhilesh-Saggere** [comment]
Hi @yesoreyeram ,

I am using the infinity plugin to fetch the milestone data from my gitlab instance on grafana.

Example data frame: 

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ALEU1C~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ALEU1C~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


Milestone | due_date | Deadline
-- | -- | --
M1 | 16/07/2024 | 3 months ago
M2 | 25/11/2024 | in a month
M3 | 15/12/2024 | in 3 months
M4 | 16/01/2025 | in 4 months



</body>

</html>

I want to conditionally highlight the milestone which is nearing (say in a month's time)
And want to highlight in red for the milestone that has passed.

I have tried using the field override option, (to set threshold/colour mapping) but the rows aren't getting changed as per the expectation.
I think this is because the deadline column is not a string but the time value. (which is created by the filed override by comparing it with today's time - by following https://community.grafana.com/t/add-timestamp-column-to-infinity-datasource/116761)

How can i achieve this functionality and have good visualisation in dashboard.
I also tried to duplicate the coloums and change the format, it looks to be tough. 
Are there any possiblities to add gnerate coloumns (like measured tables or DAX queries incase of a power BI).

I am finding it hard to figure this out, could you please share your thoughts or point me to the resources that would help me to understand and solve this problem.

Thanks in advance.

**Comment by thejas2511** [comment]
@yesoreyeram  Hey , i am trying to create an alert based on a health API in my service, I want to send an alert when my api response status is 5XX, but infinity plugin is throwing error when API fails, is there a way to catch the error and send out an alert.

**Reply by yesoreyeram** [reply]
Nope. Not at the moment. You can use grafana synthetic monitoring/k6 instead.

**Comment by Narcolapser** [comment]
@yesoreyeram 

I did the example at the top. I have data in a similar format so I was excited to see the example. But the section with the columns made me sad. In my particular data the fields are not known ahead of time. If I choose as my panel type "table" the data shows up perfectly, including clear time stamp, but if I use "time series" it just says "Data is missing a time field". I do this sort of thing all the time with Influx, I'd rather not export my API data to influx just so I can read it in there and get dynamically selected fields.

**Reply by Narcolapser** [reply]
@yesoreyeram Any thoughts? I'm very interested in being able to do this.

**Reply by yosiasz** [reply]
would be helpful to see sample data

**Reply by Narcolapser** [reply]
I used the data in the original post at the top. To be clear: https://thingspeak.com/channels/38629/feed.json

**Reply by yosiasz** [reply]
![image](https://github.com/user-attachments/assets/d379e801-4e57-43c1-88c8-9d4278ca15da)

**Reply by yosiasz** [reply]
or

![image](https://github.com/user-attachments/assets/365ed529-be05-42a3-bc57-9eda6e4ae1e0)

**Reply by Narcolapser** [reply]
ok, in both of these cases you are explicitly specifying the columns. Either with:

```
parse-json
| jsonata "$.feeds.{'time': created_at, 'value': field1}"
```

or with: 

<img width="503" alt="image" src="https://github.com/user-attachments/assets/d4e1f182-9ca6-4f58-968a-c83535bfd967" />

Is there a way to have these columns be automatically selected and still have a timestamp? In the table format it clearly dynamically knows which columns are there and the created_at is clearly a timestamp:

<img width="1277" alt="image" src="https://github.com/user-attachments/assets/933d8d0a-3875-4f76-bef2-2915cfbdf027" />

Do you know of any way to both have time series data and have the columns show up dynamically so that I don't have to update my dashboard every time my underlying data adds a new field. Thanks for your assistance!

**Reply by yosiasz** [reply]
please provide example of such situation where columns change. remember time series requires, afaik, these named columns: `time,value,metric`

**Reply by Narcolapser** [reply]
Ok, i created a highly contrived example here: https://example.studiosleepygiraffe.com/example.json

The columns change every time you reload it. I've got a couple mapped out with this query:

```
parse-json
| jsonata "$.{'time': timestamp, 'current': current, 'wind_speed': wind_speed, 'altitude': altitude, 'humidity': humidity, 'temperature': temperature}"
```

It successfully produces a graph like so:

<img width="1595" alt="image" src="https://github.com/user-attachments/assets/61233277-3891-4e28-97eb-9fabe9a73f49" />

But only the columns that are in my query show up, you can see in the code [here](https://github.com/Narcolapser/quickexample/blob/master/main.py) that there is still "voltage" and "pressure" that I haven't mapped. I'll admit this is a highly contrived example, but it is for demonstration purposes. My real example is a time tracking program I use which provides a json end point which I can read into grafana. Every now and then the categories in my time tracking change, typically me adding one. Just the other day I added a new category for time spent with our intern. Suddenly my graphs showed me not working for many hours of the day. Once I added "intern" as a column my graphs looked good again, but I don't want to have to do that. I want it to automatically detect that new column. So this example should work well as a demonstration of that idea. You could imagine this being the data coming off of a weather station as someone adds and removes instruments.

**Reply by yosiasz** [reply]
a viable option is for you to use python to also regenenate the grafana dashboard with the new columns, or modify it's json content and add the new columns/removed columns. sort of a provisioning by code method. Also look into grafana dashboard api for that also.

**Reply by yosiasz** [reply]
or maybe you could use jsonata Functional Programming to iterate on fields and dynamically process them hmmmmm

https://docs.jsonata.org/programming

or this

https://try.jsonata.org/eC3edExGI

the extra metric foo comes through

![image](https://github.com/user-attachments/assets/b3f66123-a501-4989-8b34-eb40684a35d7)

run with it 🏇
